Add polynomial interpolation and ability to read RotNS data

nilsdeppe opened the pull request:
## Proposed changes

- Add common polynomial interpolation code where the degree is a template parameter to eliminate memory allocations
- Minor cleanups related to clang-tidy
- Add ability to read RotNS data. This is created by the RotNS code described in the Cook, Shapiro, Teukolsky 92 and 94 papers. It allows for constructing equilibrium differentially and uniformly rotating NSs. In the future we will probably want to add density perturbations and magnetic fields, but those technically will require solving the constraint equations for the new metric, in which case having the elliptic solver just create the full initial data would be nicer.
- Add GRMHD executable for rotating NS. While this technically is just hydro, we don't have subcell for hydro only, and ultimately things are much more interesting with magnetic fields...

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add ID for rotating NS for RotNS code by CST

nilsdeppe committed:
Add polynomial interpolation

nilsdeppe committed:
Update FishboneMoncriefDisk pup NOLINT style

nilsdeppe committed:
ClangTidy update for Clang-13

nilsdeppe committed:
Add EvolveValenciaDivCleanRotatingStar executable

wthrowe commented on src/NumericalAlgorithms/Interpolation/PolynomialInterpolation.hpp:
```diff
@@ -0,0 +1,21 @@
+ */
+template <size_t Degree>
+void polynomial_interpolation(gsl::not_null<double*> y,
+                              gsl::not_null<double*> error_in_y,
```
Document `error_in_y`.

wthrowe commented on src/NumericalAlgorithms/Interpolation/PolynomialInterpolation.cpp:
```diff
@@ -0,0 +1,77 @@
+      const double c_minus_d =
+          gsl::at(algorithm_c, i + 1) - gsl::at(algorithm_d, i);
+      ASSERT(delta_x_o - delta_x_p != 0.0,
+             "Uniformly spaced grid points or repeated grid points. i:"
```
If uniformly spaced points don't work that should be documented.  (Repeated is kind of obvious.)

wthrowe commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_PolynomialInterpolation.cpp:
```diff
@@ -0,0 +1,44 @@
+      make_not_null(&target_y), make_not_null(&error_in_y), target_x,
+      gsl::make_span(y.data(), Degree + 1),
+      gsl::make_span(coords.data(), Degree + 1));
+  CHECK_ITERABLE_APPROX(target_y, expected_y);
```
Check `error_in_y`.

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp:
```diff
@@ -0,0 +1,957 @@
+  int num_angular_points = -1;
+  cst_file >> num_radial_points >> num_angular_points;
+  num_radial_points_ = static_cast<size_t>(num_radial_points);
+  num_angular_points_ = static_cast<size_t>(num_angular_points) + 1;
```
I think this line deserves a comment.  Why is it one more than the number read from the file?

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp:
```diff
@@ -0,0 +1,957 @@
+    // radius_index is an estimate it could be that you compute an index that's
+    // slightly an over-estimate. We could do a binary search to find the exact
+    // index if the estimate fails.
+    // radius_index--;
```
Leaving this code commented is fine if we are unsure if it is needed, but comment the whole section (including the `if`).

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.hpp:
```diff
@@ -0,0 +1,574 @@
+ *   u_z&=0.
+ * \f}
+ *
+ * \warning Near (within `1e-2`) $r=0$ the numerical errors from interpolation
```
`\f$r=0\f$`

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp:
```diff
@@ -0,0 +1,957 @@
+
+template <typename DataType>
+auto RotatingStar::make_metric_data(const size_t num_points) const ->
+    typename IntermediateVariables<DataType>::MetricData {
```
Could this be a constructor for `MetricData`?

wthrowe reviewed (changes requested):
I did not go through all the math in detail, but I expect any errors there would have been pretty obvious when the code was used.

nilsdeppe commented:
rebase and pushed fixup commits. Thanks for the review @wthrowe!

wthrowe commented:
Looks good.  Squash.

nilsdeppe commented:
Squashed! Thanks!

mduez commented:
I am confused by this comment in the documentation:
"where the rotation is about the y-axis. We switch to rotation about the z-axis while switching to Cartesian components below."
RotNS does not, I believe, invoke Cartesian coordinates at all.  SpEC does take the x-axis as the rotation axis by default, but there is nothing imposing rotation axis = y in the RotNS output that needs to be rotated.  Can't you just say that your code takes the rotation axis to be the z-axis?

mduez commented:
I believe your definition of v is the Eulerian velocity rather than the transport velocity.  If that's standard in SpECTRE, there's no need to note this.  In SpEC, v is the transport velocity, and we need to be careful to specify which velocity we mean--especially when it's a different one.

nilsdeppe commented:
Rebased and squashed in rewording to "We use rotation about the \f$z\f$-axis." 

We use `v^i` to be the Eulerian velocity. I forget what exactly we use for the transport velocity (`v^i_t` maybe), but they're distinct.

kidder commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.hpp:
```diff
@@ -0,0 +1,576 @@
+ *   \begin{pmatrix}
+ *     \cos(\phi) \sin(\theta) & \sin(\theta)\sin(\phi) & \cos(\theta) \\
+ *     \tfrac{\cos(\phi)\cos(\theta)}{r} & \tfrac{\sin(\phi)\cos(\theta)}{r}
+ *     & \tfrac{\cos(\phi)}{r} \\
```
should be `-sin(theta)/r`

kidder commented on src/NumericalAlgorithms/Interpolation/PolynomialInterpolation.cpp:
```diff
@@ -0,0 +1,77 @@
+  // intermediate variable to make life easier.
+  std::array<double, order> algorithm_c{};
+  std::array<double, order> algorithm_d{};
+  const size_t num_points = x_values.size();
```
you assert above that this is equal to `order` so why introduce a new variable?

kidder commented on docs/References.bib:
```diff
@@ -448,6 +448,32 @@ @book{Cockburn1999
+  doi =          {10.1086/171849},
+}
+
+@ARTICLE{Cook1994,
```
these aren't generating links to the papers

nilsdeppe replied:
I don't know of any online verison of these papers

kidder commented on src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp:
```diff
@@ -0,0 +1,957 @@
+  }
+  for (size_t i = 0; i < num_points; ++i) {
+    if (get_element(*radius, i) < 1.e-12) {
+      get_element(*cos_theta, i) = 0.5;
```
why is this theta chosen?

kidder reviewed (changes requested):
you can squash

nilsdeppe commented:
rebased and squashed. Thanks for the reviews!
